javalib/src/main/scala/java/util/concurrent/ConcurrentHashMap.scala

```diff
@@ -0,0 +1,143 @@
+      remove(key)
+      true
+    } else false
+
+  def replace(key: K, value: V): V =
+    if (containsKey(key)) {
+      put(key, value)
+    } else get(key)
+
+  def replace(key: K, oldValue: V, newValue: V): Boolean =
+    if (containsKey(key) && get(key).equals(oldValue)) {
+      put(key, newValue)
+      true
+    } else false
+
+}
+
+class ConcurrentHashMap[K >: Null, V >: Null](initialCapacity: Int,
+    loadFactor: Float,
+    concurrencyLevel: Int,
+    initialMap: java.util.Map[_ <: K, _ <: V])
+    extends java.util.AbstractMap[K, V]
+    with java.util.concurrent.ConcurrentMap[K, V]
+    with java.io.Serializable {
+
+  def this() = 
+    this(16, 0.75F, 16, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialCapacity: Int) = 
+    this(initialCapacity, 0.75F, 16, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialCapacity: Int, loadFactor: Float) = 
+    this(initialCapacity, loadFactor, 16, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialCapacity: Int, loadFactor: Float, concurrencyLevel: Int) = 
+    this(initialCapacity, loadFactor, concurrencyLevel, mapAsJavaMap(Map[K,V]()))
+
+  def this(initialMap: java.util.Map[_ <: K, _ <: V]) = 
+    this(16, 0.75F, 16, initialMap)
+
+  putAll(initialMap)
+
+  private lazy val inner = new scala.collection.mutable.HashMap[K, V]()
+
+  override def clear(): Unit =
+    inner.clear
+
+  override def containsKey(key: Any): Boolean =
+      inner.contains(key.asInstanceOf[K])
+    
+  override def containsValue(value: Any): Boolean =
+    inner.exists { case (ik, iv) => value.equals(iv) }
+    
+  def elements(): java.util.Enumeration[V] =
+    asJavaEnumeration(inner.values.iterator)
+
+  def entrySet(): java.util.Set[java.util.Map.Entry[K, V]] =
+    setAsJavaSet(
+        inner.map { case (k, v) => new java.util.AbstractMap.SimpleImmutableEntry(k, v) }.toSet)
+
+  override def get(key: Any): V =
+    inner.get(key.asInstanceOf[K]).getOrElse(null)
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  def keys(): java.util.Enumeration[K] =
+    asJavaEnumeration(inner.keys.iterator)
+
+  // This will not work if used with JDK 8
+  //override def keySet(): java.util.Set[K] =
+  //	setAsJavaSet(inner.keys.toSet)
+
+  override def put(key: K, value: V): V =
+    try {
+      inner.put(key, value)
+      value
+    } catch {
+      case _: Throwable => null
+    }
+
+  override def putAll[K1 <: K, V1 <: V](m: java.util.Map[K1, V1]): Unit =
+    for {
+      e <- asScalaSet(m.entrySet())
+    } yield {
+      put(e.getKey, e.getValue)
+    }
+
+  override def putIfAbsent(key: K, value: V): V =
+    inner.getOrElseUpdate(key, value)
+    
+  override def remove(key: Any): V =
+    inner.remove(key.asInstanceOf[K]).get
+
+  override def remove(key: Any, value: Any): Boolean =
+    try {
+      val old = inner(key.asInstanceOf[K])
+      if (value.equals(old)) {
+        inner.remove(key.asInstanceOf[K]).isDefined
+      } else false
+    } catch {
+      case _: Throwable => false
```

Comment:
Actually, I don't understand why you catch anything at all (and again, same for the two methods below).